### PR TITLE
Add replay recording support

### PR DESCRIPTION
**Pull request by frenzibyte**
Currently has to convert from and to legacy replay frames due to no modern replay encoding support osu!lazer-side yet, which limits tertiary and quaternary actions from being saved as-is in replays, but looks to be working as expected to for now.

### CODE REVIEW

**Comment by LumpBloom7** [comment]
Seeing as rush doesn't record any positional inputs, it would be possible to repurpose one of those values as a bitflag of some sort, similar to mania's implementation. This will make sure the playback matches the actions performed during recording.

**Comment by frenzibyte** [comment]
Yeah now I realize what you meant back then, will see.

File: osu.Game.Rulesets.Rush/Replays/RushReplayFrame.cs [hunk 1]
```diff
@@ -41,49 +41,51 @@ public void FromLegacy(LegacyReplayFrame currentFrame, IBeatmap beatmap, ReplayF
...
+            if (actions.Contains(RushAction.GroundPrimary)) flags |= RushActionFlags.GroundPrimary;
+            if (actions.Contains(RushAction.GroundSecondary)) flags |= RushActionFlags.GroundSecondary;
+            if (actions.Contains(RushAction.GroundTertiary)) flags |= RushActionFlags.GroundTertiary;
+            if (actions.Contains(RushAction.GroundQuaternary)) flags |= RushActionFlags.GroundQuaternary;
 
             return flags;
         }
 
-        private static IEnumerable<RushAction> getActionsFromFlags(RushSideActionFlags sideFlags)
+        private static IEnumerable<RushAction> getActionsFromFlags(RushActionFlags flags)
```
Comment (swoolcock) [inline osu.Game.Rulesets.Rush/Replays/RushReplayFrame.cs:64]: Likewise for this:
```cs
private static IEnumerable<RushAction> getActionsFromFlags(RushActionFlags flags)
{
    RushActionFlags flag;
    int action = 0;
    while ((flag = (RushActionFlags)(1 << action++)) <= RushActionFlags.LastAction)
    {
        if (flags.HasFlagFast(flag))
            yield return (RushAction)flag;
    }
}
```

If you add to `RushActionFlags` (please do this anyway):

```cs
LastAction = GroundQuaternary
```

**Review by swoolcock — CHANGES_REQUESTED** [review]

File: osu.Game.Rulesets.Rush/Replays/RushReplayFrame.cs [hunk 2]
```diff
@@ -41,49 +41,51 @@ public void FromLegacy(LegacyReplayFrame currentFrame, IBeatmap beatmap, ReplayF
...
-            if (Actions.Contains(RushAction.AirSecondary)) state |= ReplayButtonState.Left2;
-            if (Actions.Contains(RushAction.GroundPrimary)) state |= ReplayButtonState.Right1;
-            if (Actions.Contains(RushAction.GroundSecondary)) state |= ReplayButtonState.Right2;
-
-            return new LegacyReplayFrame(Time, (float)getFlagsFromActions(Actions), 0f, state);
-        }
+        public LegacyReplayFrame ToLegacy(IBeatmap beatmap) => new LegacyReplayFrame(Time, (float)getFlagsFromActions(Actions), 0f, ReplayButtonState.None);
 
-        private static RushSideActionFlags getFlagsFromActions(IEnumerable<RushAction> actions)
+        private static RushActionFlags getFlagsFromActions(IEnumerable<RushAction> actions)
```
[the remark this answers is not in the document]
Comment (frenzibyte) [reply osu.Game.Rulesets.Rush/Replays/RushReplayFrame.cs:48]: Second one sounds too much there, let's use the first one

**Pull request merged**